src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs

```diff
@@ -981,13 +984,24 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
                             {
                                 diagnostics.Add(ErrorCode.ERR_CantChangeRefReturnOnOverride, overridingMemberLocation, overridingMember, overriddenMember);
                             }
-                            else if (!overridingMethod.ReturnTypeWithAnnotations.Equals(overriddenMethod.ReturnTypeWithAnnotations, TypeCompareKind.AllIgnoreOptions))
+                            else if (!IsValidOverrideReturnType(overridingMethod, overridingMethod.ReturnTypeWithAnnotations, overriddenMethod.ReturnTypeWithAnnotations, diagnostics))
                             {
                                 // if the Return type is or contains an error type, the return type must be fixed before the override can be found, so suppress error
                                 if (!isOrContainsErrorType(overridingMethod.ReturnType))
                                 {
-                                    // error CS0508: return type must be 'C<V>' to match overridden member 'M<T>()'
-                                    diagnostics.Add(ErrorCode.ERR_CantChangeReturnTypeOnOverride, overridingMemberLocation, overridingMember, overriddenMember, overriddenMethod.ReturnType);
+                                    // If the return type would be a valid covariant return, suggest using covariant return feature.
+                                    HashSet<DiagnosticInfo> discardedUseSiteDiagnostics = null;
+                                    if (DeclaringCompilation.Conversions.HasIdentityOrImplicitReferenceConversion(overridingMethod.ReturnTypeWithAnnotations.Type, overriddenMethod.ReturnTypeWithAnnotations.Type, ref discardedUseSiteDiagnostics))
```

Comment:
This seems strange. Shouldn't we check if it's valid, including a covariant return, then check if the feature is enabled and, if it isn't and the two types are not equal, produce an error?

Reply:
I'd rather keep the error case separate. rather than blend the code for the correct case and the error case.

---
In reply to: [414069899](https://github.com/dotnet/roslyn/pull/43576#discussion_r414069899) [](ancestors = 414069899)